### PR TITLE
[OidcClient] add a method to fetch introspection URL

### PR DESCRIPTION
**Pull request by David-Dadon**
Our company need to fetch information from the introspection

### CODE REVIEW

File: src/OidcClient.php [hunk 1]
```diff
@@ -238,6 +238,33 @@ public function retrieveUserInfo(OidcTokens $tokens): OidcUserData
...
+
+    // Read the data
+    $data = json_decode($jsonData, true);
+
+    // Check data due
+    if (!is_array($data)) {
+      throw new OidcException('Error from the introspection endpoint.');
+    }
+
+    return $data;
```
Comment (bobvandevijver) [inline src/OidcClient.php:265]: I'd like to see a model for this, as the response from the introspection endpoint is strictly defined: https://datatracker.ietf.org/doc/html/rfc7662#section-2.2. 

Best is something similar to the OidcUserData model.

File: src/OidcClientInterface.php [hunk 2]
```diff
@@ -82,4 +82,9 @@ public function generateEndSessionEndpointRedirect(
    * @throws OidcException
    */
   public function retrieveUserInfo(OidcTokens $tokens): OidcUserData;
+
+  /**
+   * @throws OidcException
+   */
```
Comment (bobvandevijver) [inline src/OidcClientInterface.php:88]:
```suggestion
  /**
   * Introspect the supplied token
   *
   * @throws OidcException
   */
```

File: src/OidcClient.php [hunk 3]
```diff
@@ -238,6 +238,33 @@ public function retrieveUserInfo(OidcTokens $tokens): OidcUserData
...
+  {
+    $headers = [];
+    if (in_array('client_secret_basic', $this->getTokenEndpointAuthMethods())) {
+      $headers = [
+        'Authorization: Basic ' . base64_encode(urlencode($this->clientId) . ':' . urlencode($this->clientSecret)),
+      ];
+    }
+
+    $jsonData = $this->urlFetcher->fetchUrl(
+      $this->getConfigurationValue('introspection_endpoint'),
```
Comment (bobvandevijver) [inline src/OidcClient.php:251]: Please create a `getIntrospectionEndpoint` method to be consistent with the other endpoint retrievals.

**Review by bobvandevijver — CHANGES_REQUESTED** [review]
Thank you for your MR and work so far! I have made some comments so it can make a great and complete addition for this library. Do you think you can work with them?

**Comment by David-Dadon** [comment]
Hi, I made some corrections after your review of my MR
Best regards

File: src/OidcClient.php [hunk 4]
```diff
@@ -238,6 +240,43 @@ public function retrieveUserInfo(OidcTokens $tokens): OidcUserData
...
+    } elseif ($tokenType === OidcTokenType::REFRESH) {
+      $params = [
+        'token' => $tokens->getRefreshToken(),
+        'token_type_hint' => 'refresh_token',
+      ];
+    } else {
+      $params = [
+        'token' => $tokens->getIdToken(),
+      ];
+    }
```
Comment (bobvandevijver) [inline src/OidcClient.php:264]:
```suggestion
    $params = match ($tokenType) {
      OidcTokenType::ACCESS => [
        'token' => $tokens->getAccessToken(),
        'token_type_hint' => 'access_token',
      ],
      OidcTokenType::REFRESH => [
        'token' => $tokens->getRefreshToken(),
        'token_type_hint' => 'refresh_token',
      ],
      default => throw new InvalidArgumentException('Only access and refresh tokens can be introspected'),
    }
```

**Review by bobvandevijver — APPROVED** [review]

**Comment by bobvandevijver** [comment]
Merged with 3ab815a166cc4872d308e1c5f45727905520ded6, thank you!

**Comment by David-Dadon** [comment]
So fast, thank so much, also for your advices

**Pull request closed without merging**